src/dag.rs

```diff
@@ -995,3 +1027,124 @@ impl<D: DagLike, S: SharingTracker<D>> Iterator for PreOrderIter<D, S> {
         None
     }
 }
+
+/// Iterates over a DAG in _pre order_, yielding extra state changes.
+///
+/// Unlike the regular pre-order iterator, this one may yield each node
+/// multiple times: once in pre-order and then again after each of its
+/// children has been yielded. The different cases can be distinguished
+/// by inspecting the [`PreOrderIterItem::n_children_yielded`] and
+/// [`PreOrderIterItem::is_complete`] fields of the yielded item.
```

Comment:
It might be clearer to call this a "verbose pre-order" where first the parent, then the left child, then the parent, then right child, and finally the parent are yielded.